src/argilla/server/apis/v1/handlers/datasets.py

```diff
@@ -210,24 +211,70 @@ async def _get_search_responses(
     parsed_metadata: List[MetadataParsedQueryParam],
     limit: int,
     offset: int,
-    query: Optional["StringQuery"] = None,
+    text_query: Optional["TextQuery"] = None,
+    vector_query: Optional["VectorQuery"] = None,
     user: Optional[User] = None,
     response_statuses: Optional[List[ResponseStatusFilter]] = None,
     sort_by_query_param: Optional[Dict[str, str]] = None,
 ) -> "SearchResponses":
+    vector_settings = None
+    record = None
+
+    if vector_query:
+        if vector_query.record_id is not None:
+            record = await datasets.get_record_by_id(db, vector_query.record_id)
```

Comment:
You need to get the record by `id` and `dataset_id` if not you could potentially search by any `record_id` (even those outside of the dataset).

An alternative to this is to add/use a policy.

A good idea to check this is to add a new test trying to find by a record belonging to a different dataset and see if it's failing. After that fix the problem and make the test passing.